src/bal/dx12/src/native/com.rs

```diff
@@ -0,0 +1,101 @@
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::ops::Deref;
+use std::ptr;
+use winapi::um::unknwnbase::IUnknown;
+use winapi::Interface;
+use D3DResult;
+
+pub struct WeakPtr<T>(*mut T);
+
+impl<T> WeakPtr<T> {
+    pub fn null() -> Self {
+        WeakPtr(ptr::null_mut())
+    }
+
+    pub fn from_raw(ptr: *mut T) -> WeakPtr<T> {
+        WeakPtr(ptr)
+    }
+
+    pub fn is_null(&self) -> bool {
+        self.0.is_null()
+    }
+
+    pub fn as_ptr(&self) -> *const T {
+        self.0
+    }
+
+    pub fn as_mut_ptr(&self) -> *mut T {
+        self.0
+    }
+}
+
+impl<T: Interface> WeakPtr<T> {
+    pub fn as_unknown(&self) -> &IUnknown {
```

Comment:
This function must be marked `unsafe`.
Otherwise this "safe" expression `WeakPtr::from_raw(1usize as *mut u32).as_unknown()` will trigger UB.